IcingaDB: better config and state update queueing

yhabteab opened the pull request:
This pull request introduces a new runtime changes queue to IcingaDB, along with several enhancements to the `RedisConnection` class. These changes aim to improve the memory footprint and number of duplicate (and thus superfluous) Redis queries. The problem of duplicate queries has been a long-standing issue in IcingaDB, and some hacky workarounds have been implemented in the past to mitigate it. This PR takes a more systematic approach as Julian described in #10186 to address the root cause. I will try to summarize the key changes below:

## Changes Queue

A new config and state update queue has been introduced to IcingaDB, which allows to batch any config or state related runtime updates for a given object in an efficient manner. This new queue works as outlined below:

First of all, this new way of queueing is only used for a specific set of updates, which include but are not limited to:

+ Object creation and deletion
+ Object modification (e.g., updating attributes)
+ Relationship updates (e.g., adding or removing links between objects)
+ State related updates (e.g., state changes, acknowledgements, etc.)
+ Dependency config and state updates

This new queue **does not** cover any history related writes, those types of events follow a different path and are not affected by this change. The focus here is solely on runtime object changes that affect the normal non-historical operation of IcingaDB. Consequently, history and heartbeat related writes use their own dedicated Redis connection and do not interfere with any of the changes described here.

Now, here is how this queue operates:

When an object is modified, instead of immediately writing the changes to Redis, the object pointer is pushed onto the queue with a corresponding flag indicating the type of change required. As long as the object remains in the queue, any subsequent Redis write requests concerning that object are merged into the existing queued dirty bits. This means that no matter how many times e.g., a `OnStateChange` is triggered for a given object, only a single write operation will be performed when it is finally popped from the queue. Do note that an object can have multiple dirty bits set, so if both its attributes and state are modified while in the queue, a state and config update will be sent when it is processed.

The consumer of this queue is a new background worker that pops objects from the queue and performs the necessary Redis write operations. This worker doesn't immediately process objects as they are enqueued; instead, it waits for a short period (currently set to `300ms`) to allow for more changes to accumulate and be merged. After this wait period, the worker serializes the queued objects according to their dirty bits and sends the appropriate Redis commands. Though, there's also another restriction in place: when the used `RedisConnection` reaches a certain number of pending commands (currently set to `128`), the worker won't dequeue any more objects from the queue until the pending commands drop below that threshold. This ensures that we don't unnecessarily waste memory by serializing too many objects in advance, if the Redis server isn't able to keep up.

To accommodate this new changes queue, quite a number of existing code has been refactored, so that we no longer perform immediate writes to Redis. Additionally, the `RedisConnection` class has been enhanced to support this new workflow.

## RedisConnection Enhancements

Several enhancements have been made to the `RedisConnection` class to better support the changes queue and improve overall efficiency:

1. As a consequence of the changes queue, the Redis queries internal queue has significantly been simplified. As opposed to the previous version which used a complex data structure to correctly manage the query priorities, this version uses two separate `std::queue`s to manage the pending queries. One queue is used for queries with high priority, while the other queue is used for low priority queries. High priority queries have always, well, higher priority than low priority queries, and can overtake them at any time. This new approach is much simpler and more efficient, as it eliminates the need for complex data structures and allows for straightforward management of query priorities.
2. And while I'm at it, I also took the chance to improve the `WriteQueueItem` type by replacing the previously used ridiculously verbose query types by a more compact `std::variant` based approach. This not only reduces memory usage but also makes clearer that each item represents exactly one of a defined set of query types and nothing else.

Now, IcingaDB is subscribed to the `OnNextCheckChanged` signal and not the dummy `OnNextCheckUpdated` signal anymore. Though, that dummy signal is still there since the IDO relies on it. The only behavioral change in IcingaDB as opposed to before is that the oldest pending Redis query is determined only on the primary Redis connection (the one used for history and heartbeats). If you guys think this is a problem, I can look into a way to have IcingaDB consider all connections when determining the oldest pending query.

resolves #10186

yhabteab committed:
IcingaDB: use a separate Redis conn for history & heartbeat streams

yhabteab committed:
IcingaDB: track object changesets efficiently

yhabteab committed:
IcingaDB: enqueue config runtime updates to the worker queue

yhabteab committed:
Revert "IcingaDB: suppress state sync until config sync finished"

This reverts commit f6f7d9b63508f0e3f0c0a3e31fedc4dbd8c4b85a and all
other its new users.

yhabteab committed:
RedisConnection: simplify query prioritization logic

As opposed to the previous version which used a complex data structure
to correctly manage the query priorities, this version uses two separate
queues for the high and normal priority writes. All high priority writes
are processed in FIFO order but over take all queries from the normal
priority queue. The later queue only be processed when the high priority
queue is empty.

yhabteab committed:
RedisConnection: enhance `WriteQueueItem` & related usages

yhabteab committed:
Revert "CheckerComponent#CheckThreadProc(): also propagate next check update to Icinga DB"

This reverts commit e9b8c679758bfc834085d2b878d2ac04c2122adf.

yhabteab committed:
IcingaDB: subscribe to `OnNextCheckChanged` signal

We can't drop the `OnNextCheckUpdated` signal entirely yet, as IDO still
relies on it.

jschmidt-icinga commented on lib/icingadb/redisconnection.cpp:
```diff
@@ -108,23 +138,39 @@ void LogQuery(RedisConnection::Query& query, Log& msg)
+			m_Queues.PushFront(std::move(qitem));
+		} else {
+			m_Queues.Writes.push_back(std::move(qitem));
+		}
```
This whole pushing-logic is used almost verbatim in two places. Since we have the `PushFront()` method on the queue, you might want to unify this into a single `Push()` function with a priority parameter.

yhabteab replied:
Done.

jschmidt-icinga commented on lib/icingadb/redisconnection.hpp:
```diff
@@ -218,48 +201,48 @@ namespace icinga
-		String m_CipherList;
-		double m_ConnectTimeout;
-		DebugInfo m_DebugInfo;
+		ObjectImpl<IcingaDB>::ConstPtr m_IcingaDB; // The IcingaDB object this connection belongs to.
```
This introduces some tight bidirectional coupling between the `IcingaDB` (even though it's just the `-ti` file here) and `RedisConnection` objects. It would be nice if this could be avoided, though copying all the members doesn't seem elegant either.

jschmidt-icinga replied:
I had already thought about it and couldn't come up with anything good to suggest, other than designing it that way from the start. One option would be to maybe make the members `string_view`s, which would be fine here, since the connection object has a longer lifetime than the icingadb object, but might become an implementation detail when/if other classes want to use this class. Or just suck up the (small) memory cost and leave things as they were.

It's not the worst thing in the world either way. We have this kind of coupling in many places, like `(JsonRpc|HttpServer)Connection`<->`ApiListener`, but it is kind a ugly, design-wise.

Regarding `RedisConnection` being only used by IcingaDB: Recently I was briefly looking at caching Perfdata in Redis for persistence in case the target services go offline. It's always nice to at least keep the option open of using a class like that for something else in the future.

jschmidt-icinga commented on lib/icingadb/redisconnection.hpp:
```diff
@@ -218,48 +201,48 @@ namespace icinga
 			// Requestors, each waiting for multiple responses at once
 			std::queue<std::promise<Replies>> RepliesPromises;
 			// Metadata about all of the above
 			std::queue<FutureResponseAction> FutureResponseActions;
```
Is there any reason these need to live together in the same struct? Since you don't touch the code that deals with the Replies much, that can stay as it is, but I'd really remove the `Writes` queue from this struct and make it its own thing (ideally a class with only `Push()` `Pop()` methods).

yhabteab replied:
> Edit: Case in point, while reading the code it made me look up what the other queues do and in what way they relate to the write queue. It implies a relationship that doesn't exist.

There does exist a some sort of relationship though. I mean, if you don't send anything, there won't be any responses to handle. So, I'm not sure how to proceed with this. I'll leave it as is for now and maybe can discuss it in person later.

jschmidt-icinga replied:
No, it's fine. It's just an aesthetic preference and something I personally found slightly confusing. Since this doesn't have any functional input we can just leave it the way it is.

jschmidt-icinga commented on lib/icingadb/icingadb.hpp:
```diff
@@ -27,6 +31,186 @@
+ */
+struct PendingQueueItem
+{
+	int DirtyBits;
```
Regardless what you decide above, you can use `std::underlying_type_t<icinga::DirtyBits>` here.

jschmidt-icinga reviewed (commented):
I didn't test extensively yet, but from a quick test everything seems to work as expected. I also can't speak much for the logic and performance implications of when to send which events to Redis since I barely touched that until now. I'll continue to look at this in the coming days and see if I can test this more thoroughly.

yhabteab commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,521 @@
+			for (auto it(view.begin()); it != view.end(); ++it) {
+				const PendingItemVariant itemToProcess = *it;
+				bool abort = std::visit(
+					[this, &view, &it, &lock, &itemToProcess, &now, &retryAfter, &madeProgress](const auto& item) {
```
> I suggest to at least simplify the function to an outer event-loop and an inner iterative loop. You can signal abort/madeProgress conditions by setting `it` to `view.end()`.

That definitely makes sense. Honestly, I'm a bit embarrassed that the function don't exactly what I was aiming for when I wrote it :-). The inner do-while loop is completely unnecessary, and bypasses all the preconditions given by the `canContinue` predicate. I'll fix that.

> There's also some redundant checks in the outer loop, which is probably why the middle loop was necessary.

Can you please clarify which checks you are referring to? The outer loop checks `GetActive()` with each iteration, which is necessary to determine whether the thread should continue running or exit. But it also checks `GetActive()` in the body of the loop after waking up from the condition variable wait, which might seem redundant. However, this is not actually redundant because if we are woken up from the wait, Icinga DB might be in the process of shutting down, and we don't want to proceed with processing items since we don't have a working RedisConnection anymore.

> only wait for the conditions that actually notify and check those that never change (like `m_ConfigDumpDone` only once before the loop).

Only once? While waiting on `m_ConfigDumpDone` to become true, we might still receive notifications on `m_PendingItemsCV` that new items have been enqueued. That's the reason I've guarded it with an extra while loop inside the main loop.

> That way you don't need to copy the items (which incurs multiple ref-count increments) at all, instead of for each iteration as it currently is.

That copy is not unintentional though, once the worker thread picks an item to process, it **must** be removed from the container while holding the lock (and if you erase something from the container, you need a copy of the item first). Otherwise, another thread might mutate and even completely erase the item we're about to process right after we released the lock. I've not looked yet into your changes, but if the motivation is to avoid ref-count increments, then that won't work out. Though, looking at the code again, I don't have to always copy the item but only when we're actually going to process it. I'll adjust that.

jschmidt-icinga replied:
>Can you please clarify which checks you are referring to?

Mostly the Abort conditions that are also part of the `while (!canContinue())` condition and then immediately rechecked in the `if ... break`.

If you look at my linked example (not sure if this is entirely correct), you'll see that I've tried to separate wait conditions (at the bottom of the loop) with abort conditions (the loop condition).

Not that this hurts a lot, I'd rather have too many checks than too few, but we should at least think about where it makes sense to check what. I'm certain your next iteration with the outer loops merged will be fine.

>Only once?

Or repeatedly as a precondition to the loop, but I meant in one `condition_variable::wait()` predicate and before the loop, not on each iteration.

>Otherwise, another thread might mutate and even completely erase the item we're about to process right after we released the lock.

Now I get it.

>I don't have to always copy the item but only when we're actually going to process it.

Seems reasonable.

yhabteab commented on lib/icingadb/icingadb-objects.cpp:
```diff
@@ -45,8 +46,7 @@
 
-using Prio = RedisConnection::QueryPriority;
-
+std::map<Type::Ptr, RedisKey> IcingaDB::m_CustomVarKeys;
```
Done.

yhabteab commented:
> I'll continue to look at this in the coming days and see if I can test this more thoroughly.

At the moment the integration tests from the [Icinga DB](https://github.com/Icinga/icingadb/blob/main/.github/workflows/integration-tests.yml) repository are the only way to stress test this PR thoroughly. I've been running them ever since the initial implementation and I have almost gone crazy due to a subtle race condition that only showed up when running those tests.

julianbrost commented on lib/icingadb/icingadb.hpp:
```diff
@@ -27,6 +31,186 @@
+ *
+ * @ingroup icingadb
+ */
+enum DirtyBits : uint
```
I don't know how you've discovered `uint`, but it's not a standard C++ type and in my system headers it's defined in a block with the comment `/* Old compatibility names for C types.  */`, so that doesn't really sound like something that should be used.

yhabteab replied:
It states `/* Sys V compatibility */` on my machine :).

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,531 @@
+		boost::hash_combine(seed, depGroup);
+	}
+	return "DUMMY_" + std::to_string(seed);
+}
```
All the hashing in C++ based on `size_t` isn't really meant to prevent collisions. Sure, collisions should be unlikely, but code using it must assume there are collisions. Keeping the overall function signature, I would rather just write both pointer values to the string, that's guaranteed that there won't be any collisions then.

However, I'm also question why this function even needs to exist, i.e. why you're mapping this to a string. Couldn't you also use some `std::variant` containing the different pointer types as the ID (or lookup key) instead?

julianbrost commented on lib/icingadb/icingadb.hpp:
```diff
@@ -242,13 +409,32 @@ class IcingaDB : public ObjectImpl<IcingaDB>
+	 * This connection is a child of @c m_HistoryCon, meaning it forwards all its connection stats
+	 * to @c m_HistoryCon as well.
+	 */
 	RedisConnection::Ptr m_Rcon;
```
> Primary Redis connection

> This connection is a child of @c m_HistoryCon

That sounds counter intuitive to me. Why is it this way around?

julianbrost reviewed (commented):
Don't consider this a full review, just what I noticed at first glance.

julianbrost commented on lib/icingadb/redisconnection.cpp:
```diff
@@ -1,85 +1,115 @@
+			waiter.async_wait(yc);
+
+			m_QueuedWrites.Set(); // Wake up write loop
+			m_QueuedReads.Set(); // Wake up read loop
```
Wait up to 5 seconds? Doesn't this just always wait 5 seconds?

And why are the read/write loops only woken after that?

yhabteab replied:
> Wait up to 5 seconds? Doesn't this just always wait 5 seconds?

Yes, it always waits 5 seconds.

> And why are the read/write loops only woken after that?

Doesn't make any difference but I've moved these above the timer now.

julianbrost commented on lib/icingadb/redisconnection.cpp:
```diff
@@ -729,3 +783,15 @@ void RedisConnection::RecordAffected(RedisConnection::QueryAffects affected, dou
+	if (m_Stopped) {
+		throw RedisDisconnected();
+	}
+}
```
`ThrowIfStopped()` might be a better name to avoid confusing it with actual asserts that aren't active in release builds and would otherwise terminate the process.

yhabteab replied:
Done.

julianbrost commented on lib/icingadb/icingadb.hpp:
```diff
@@ -27,6 +31,197 @@
+	DependencyChildRegistered = 1<<5,
+	DependencyGroupState      = 1<<6,
+
+	DeleteSubKeys = 1<<7, // Delete a sub-key in Redis (used for relationships, not associated with an object directly).
```
Where's `DeleteSubKeys` actually used? I'd expect some place in `icingadb-worker.cpp` what checks whether this bit is set.

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,509 @@
+							// other checkables but since we only track the first one that requested the update, we
+							// need to pick a random other child to trigger the update instead. It's safe to do so,
+							// because the child pointer doesn't contribute to the uniqueness of the pending item.
+							pdit.Child = depGroup->GetAnyChild();
```
Instead of tracking a child and updating it, wouldn't it make more sense to just call `GetAnyChild()` when processing the entry when it's actually needed?

yhabteab replied:
Done.

yhabteab commented on lib/icingadb/icingadb-objects.cpp:
```diff
@@ -2776,50 +2708,81 @@ void IcingaDB::SendGroupsChanged(const ConfigObject::Ptr& object, const Array::P
+		keyType = RedisKey::HostGroupMember;
+	} else {
+		keyType = RedisKey::ServiceGroupMember;
+	}
```
> 1. I'm not absolutely sure if all compilers will be happy about this, but I think inside `if constexpr` they should eliminate the other branches entirely and `static_assert(false)` should be fine. [↩](#user-content-fnref-1-7265de7f6ad1165941ba4b192da9126a)

Apparently not https://github.com/Icinga/icinga2/actions/runs/19678875589

julianbrost commented on lib/icingadb/icingadb.cpp:
```diff
@@ -81,19 +79,36 @@ void IcingaDB::Start(bool runtimeCreated)
+		GetTlsProtocolmin(),
+		GetCipherList(),
+		GetDebugInfo()
+	);
```
I'd rather initialize the field individually, i.e.:

```c++
RedisConnInfo::Ptr connInfo = new RedisConnInfo();
connInfo.EnableTls = GetEnableTls();
// ...
```

Then you can check whether the values are passed correctly here without needing IDE annotations (or very carefully checking them one by one) and also can just omit the following constructor:
https://github.com/Icinga/icinga2/blob/ac841b360c229a4d14f805ad289ac89420d72cec/lib/icingadb/redisconnection.hpp#L76-L110

julianbrost replied:
Why was the [nice initialization syntax (`{.x = 42}`)](https://en.cppreference.com/w/cpp/language/aggregate_initialization.html) only added in C++20 (when it's already in C99)? :(

Anyway, I think we could still get a reasonable level of const-correctness if there was an `RedisConnInfo::ConstPtr IcingaDB::GetRedisConnInfo()` that does the initialization and is the only place where there's any non-const use of `RedisConnInfo`.

julianbrost commented on lib/icingadb/icingadb-objects.cpp:
```diff
@@ -2115,7 +2035,7 @@ void IcingaDB::SendStartedDowntime(const Downtime::Ptr& downtime)
 	/* Update checkable state as in_downtime may have changed. */
-	UpdateState(checkable, StateUpdate::Full);
+	EnqueueConfigObject(checkable, FullState);
 
```
Never mind, as the first argument remained unchanged and wasn't highlighted in the diff, I didn't pay much attention to it. :see_no_evil:

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,503 @@
+			retryAfter = DequeueAndProcessOne(lock);
+			// Flush all remaining items as fast as possible during shutdown.
+			// At this point, there shouldn't be any new items being enqueued anymore,
+			// so we won't end up in an infinite loop.
```
What would happen if we just abandon this queue on shutdown? Is there anything of value in there? Wouldn't the next full sync overwrite everything we would potentially write here?

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,503 @@
+		// We haven't made any progress, so give it a short delay before retrying.
+		retryAfter = 10ms;
+	}
+	return retryAfter;
```
I believe that would be easier to understand if you just returned an appropriate value where you currently `break;`. For example, instead of setting `madeProgress` and the last `break;`, you could just `return 0;` (± correct type) there.

yhabteab commented on lib/icingadb/icingadb.cpp:
```diff
@@ -155,9 +151,9 @@ void IcingaDB::OnConnectedHandler()
-
+	m_ConfigDumpDone.store(true);
 	m_ConfigDumpInProgress = false;
+	m_PendingItemsCV.notify_one(); // Wake up the pending items worker to start processing items.
```
We perform all the initialization/starting stuff in the `Start` method and not in the middle of nowhere but fine I'll do it.

julianbrost replied:
For the record: this was reverted back to starting the thread in `Start()` as I later noticed that this introduces a race condition with joining the thread in `Stop()`.

julianbrost committed:
IcingaDB: use polymorphism for queue entries

julianbrost commented on lib/icingadb/icingadb.hpp:
```diff
@@ -27,6 +31,200 @@
+	RelationsKeyMap Relations;
+
+	RelationsDeletionItem(const String& id, RelationsKeyMap relations);
+};
```
As I found the overall structure a bit odd here (base class, derived classes, but then having a `std::variant` containing all the possible derived classes), I played around with this a bit. Please have a look at dc9d40fc72f6d9340495c2212147099de9aa2f0d, where I basically removed the variant and just use the class inheritance instead.

The base class now has virtual methods that are implemented by the queue item types:

https://github.com/Icinga/icinga2/blob/dc9d40fc72f6d9340495c2212147099de9aa2f0d/lib/icingadb/icingadb.hpp#L147-L158

IMHO this makes the places that used `std::visit()` more readable now and also feels like a cleaner OOP-like approach now. I'm not yet happy about the actual implementation having to be friends of the IcingaDB class, but that could be solved.

yhabteab replied:
> I don't mind, but I'm not convinced that it's an improvement.

I've cherry-picked your commit nonetheless to make progress.

julianbrost replied:
> Firstly, this is polymorphism, just not inheritance-based polymorphism, which isn't the only type.

I just found it strange to see two different mechanisms that can provide polymorphism stacked on top of each other, when one sould suffice.

If both of you prefer doing it the other way, fine for me. Though some other thing I didn't really like is the entanglement between the classes with there being variants on both the outer level, as well as within `ID` within `PendingQueueItem`, both depending on each other.

> - No additional allocation per element in the map
> - fewer cache misses due to less indirection

The second is because of the pointer in the map? Or am I overlooking yet another indirection?

> faster dispatch with std::visit than virtual dispatch

Is there really much of a difference to be expected here? In the end, both are dispatching to a function based on a dynamic value.

yhabteab commented:
I've rebased and outsourced some of the commits to a separate PR #10732 as requested and used that branch as the base for this PR.

yhabteab committed:
IcingaDB: don't send `nextupdate:{host,service}` as part of the initial sync

Don't send nextupdate as part of the intial sync. Instead, enque them to
the background worker to be sent after the initial dump is done.

yhabteab committed:
IcingaDB: enqueue all dependency related updates to the worker queue

yhabteab committed:
Fix missing `olock` for dependency child registration

Previously, the checkable was locked while processing all the dependency
registration stuff, so the worker thread should also do the same to
avoid any potential race conditions.

yhabteab committed:
IcingaDB: remove unused `UpdateObjectAttrs` method

julianbrost reviewed (commented):
See the individual comments below for a few more low-hanging fruits. I'm still going over the use of the multi index container to see if I can find a way to make this nicer.

julianbrost committed:
Simplify `IcingaDB::PendingItemsThreadProc()` event loop

yhabteab committed:
IcingaDB: put all queue related stuff into `icingadb:task_queue` namespace

julianbrost committed:
IcingaDB: use key extractor for worker queue

This commit restructures the queue items so that each one now has a method
`GetQueueLookupKey()` that is used to derive which elements of the queue are
considered to be equal. For this, there is a key extractor for the
`multi_index_container` that takes the `variant` from the queue item, calls
that method on it, and puts the result in a second variant type. The types in
that variant type are automatically deduced from the return types of the
individual methods.

yhabteab committed:
Inline `DequeueAndProcessOne` & don't process items out of order

Now, the individual `ProcessQueueItem` functions decide whether to
acquire an `olock` or not instead of probing this from within the
worker loop. This is way easier than having to deal with the potential
out of order processing of items in the queue in both ways, i.e., we
don't want to send delete events for objects while their created events
haven't been processed yet and vice versa.

yhabteab committed:
Reduce min queue item age from `1000ms` to `300ms`

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,475 @@
+				// thus we can't just process this right now when there are older items in the queue.
+				// Otherwise, we might delete something that is going to be updated/created.
+				break;
+			}
```
> Btw, now that there's also another case where we acquire an `olock` I'm not sure whether it makes sense to skip any items at all.

By second case, do you mean `PendingDependencyEdgeItem`? If so, I don't see why that would be a reason against skipping items. Skipping items also allows checkable B to be updated while checkable A is locked.

> So, I've changed the implementation to never process items out of order, but instead to always wait for some time if we can't acquire the `olock` immediately. See [006e380](https://github.com/Icinga/icinga2/commit/006e3806da1cce4e42a79176f174bbdc624c4995) for the details.

But why would the worker loop need to care about that `ObjectLock` at all anymore? Shouldn't it be fine to just leave all the locking necessary to the corresponding process function? And thereby also eliminate the following code entirely:

https://github.com/Icinga/icinga2/blob/3d7e0c4a12e2b3f885b0566526e3c9100ac3b8c9/lib/icingadb/icingadb-worker.cpp#L64-L69

yhabteab replied:
> > But why would the worker loop need to care about that `ObjectLock` at all anymore? Shouldn't it be fine to just leave all the locking necessary to the corresponding process function? And thereby also eliminate the following code entirely:
> 
> Now, that you're saying that aloud, no it doesn't need to. So, I will change that again!

Done.

julianbrost commented on lib/icingadb/icingadb-objects.cpp:
```diff
@@ -2486,27 +2410,25 @@ void IcingaDB::SendFlappingChange(const Checkable::Ptr& checkable, double change
-			Prio::CheckResult
+			}
 		);
 	}
```
Sorry for waking this up again now after quite long time. The way this is implemented now, this somewhat implies there is a case where you can't tell what should be the state in Redis.

I've took the time two write all 8 possible combinations of `checkable->GetEnableActiveChecks()`, `checkable->IsActive()` and `checkable->GetExtension("ConfigObjectDeleted")` down and found the following to be somewhat inconsistent:

Consider the situation where `checkable->IsActive() == false` and `checkable->GetExtension("ConfigObjectDeleted") == false`. Now if `checkable->GetEnableActiveChecks() == true`, nothing will be done, but if `checkable->GetEnableActiveChecks() == false`, the `ZREM` will be executed. Does that make sense? Shouldn't it either be possible to always tell whether that entry should exist in Redis and send `ZADD`/`ZREM` to ensure this, or, if this is not possible and the combination not active but also not deleted is a tricky situation (is that maybe an intermediate state during startup or shutdown where we don't want to send anything?), then do nothing in both cases?

yhabteab replied:
I don't think that this case will ever occur here but as just discussed offline, I changed the condition to `checkable->GetEnableActiveChecks() && !checkable->GetExtension("ConfigObjectDeleted")` in the first case, otherwise send `ZREM` unconditionally.

julianbrost commented on lib/icingadb/redisconnection.hpp:
```diff
@@ -38,12 +38,13 @@
 #include <memory>
 #include <queue>
-#include <set>
+#include <deque>
```
Looks like a leftover that's no longer used.

yhabteab replied:
Removed.

julianbrost commented on lib/icingadb/redisconnection.cpp:
```diff
@@ -274,29 +274,6 @@ double RedisConnection::GetOldestPendingQueryTs()
 	return future.get();
 }
```
Had a quick idea that would work in a way where the cases for the individual queues are just identical. If you like it, feel free to include it, otherwise just ignore the comment.

```c++
double RedisConnection::GetOldestPendingQueryTs() const
{
	auto promise (Shared<std::promise<double>>::Make());
	auto future (promise->get_future());

	asio::post(m_Strand, [this, promise]() {
		boost::container::static_vector<double, 2> timestamps;

		if (!m_Queues.HighWriteQ.empty()) {
			timestamps.push_back(m_Queues.HighWriteQ.front().CTime);
		}
		if (!m_Queues.NormalWriteQ.empty()) {
			timestamps.push_back(m_Queues.NormalWriteQ.front().CTime);
		}

		if (timestamps.empty()) {
			timestamps.push_back(0);
		}

		promise->set_value(*boost::range::min_element(timestamps));
	});

	future.wait();
	return future.get();
}
```

yhabteab replied:
Looks fine to me as well, so applied the changes!

julianbrost committed:
RedisConnection: simplify `GetOldestPendingQueryTs` function

julianbrost commented:
While testing, I noticed that there isn't really a way to observe how the new worker thread is performing (except seeing whether things eventually show up in Redis/Icinga Web).

<details>
<summary>Show a crude patch that adds what I wanted to know right now to IcingadbCheckTask</summary>

```diff
diff --git a/lib/icingadb/icingadb.hpp b/lib/icingadb/icingadb.hpp
index ebbaebb50..ddc87cfd6 100644
--- a/lib/icingadb/icingadb.hpp
+++ b/lib/icingadb/icingadb.hpp
@@ -484,6 +484,20 @@ private:
        void EnqueueDependencyChildRegistered(const DependencyGroup::Ptr& depGroup, const Checkable::Ptr& child);
        void EnqueueDependencyChildRemoved(const DependencyGroup::Ptr& depGroup, const std::vector<Dependency::Ptr>& dependencies, bool removeGroup);
        void EnqueueRelationsDeletion(const String& id, icingadb::task_queue::RelationsDeletionItem::RelationsKeySet relations);
+
+public:
+       std::pair<std::size_t, std::chrono::steady_clock::duration> GetPendingItemsStats()
+       {
+               std::unique_lock lock(m_PendingItemsMutex);
+
+               auto& seqView = m_PendingItems.get<1>();
+               std::chrono::steady_clock::duration maxAge{};
+               if (!seqView.empty()) {
+                       maxAge = std::chrono::steady_clock::now() - seqView.front().EnqueueTime;
+               }
+
+               return {m_PendingItems.size(), maxAge};
+       }
 };
 }
 
diff --git a/lib/icingadb/icingadbchecktask.cpp b/lib/icingadb/icingadbchecktask.cpp
index f65583116..9f21d7f15 100644
--- a/lib/icingadb/icingadbchecktask.cpp
+++ b/lib/icingadb/icingadbchecktask.cpp
@@ -183,6 +183,10 @@ void IcingadbCheckTask::ScriptFunc(const Checkable::Ptr& checkable, const CheckR
        std::ostringstream i2okmsgs, idbokmsgs, warnmsgs, critmsgs;
        Array::Ptr perfdata = new Array();
 
+       auto stats = conn->GetPendingItemsStats();
+       perfdata->Add(new PerfdataValue("icinga2_pending_items_count", double(stats.first)));
+       perfdata->Add(new PerfdataValue("icinga2_pending_items_age", std::chrono::duration_cast<std::chrono::duration<double>>(stats.second).count()));
+
        i2okmsgs << std::fixed << std::setprecision(3);
        idbokmsgs << std::fixed << std::setprecision(3);
        warnmsgs << std::fixed << std::setprecision(3);
```
</details>

That obviously needs a lot of cleanup, but should be enough what I need for testing right now. Nonetheless, it would be nice if there were some metrics that allow to verify the proper operation.

julianbrost commented:
Apart from that previous comment, I now concluded my testing without any more interesting findings, so should be mostly fine now. In particular, I also tried putting it under load and Icinga DB had no issues with that (I just managed to overload other things like the IDO and JSON-RPC :sweat_smile:).

One important use-case for having additional metrics would be that if there comes an Icinga DB related bug report, there is an easy way to check if that could be related to the changes done here. Thus, can you please add something appropriate please?

yhabteab commented:
> That obviously needs a lot of cleanup, but should be enough what I need for testing right now. Nonetheless, it would be nice if there were some metrics that allow to verify the proper operation.

Does this need to be part of the check? I would just add a periodic logging to the event loop with all the relevant metrics that can still be used without having to configure a check for it. See the history bulker for example, I would also just add a similar logging for it here.

julianbrost commented:
Would that make things much easier? Or is "can still be used without having to configure a check for it" your main motivation for that suggestion?

The `icingadb` check command already has a few `*_backlog` metrics that are pretty similar, so I'd say this would be a good fit there.

However, that doesn't mean there shouldn't be any additional logging. A warning message in case this is way behind would be helpful as well.

yhabteab committed:
Log pending items stats regularly & include them as perfdata in IcingaDB check

yhabteab commented:
> A warning message in case this is way behind would be helpful as well.

Why warning? I've just added a normal (info) logging at regular interval (always every 5m unless the worker is lagging behind) and added the the two metrics to the checkcommand's output as well.

<img width="1168" height="69" alt="Bildschirmfoto 2026-04-15 um 11 15 09" src="https://github.com/user-attachments/assets/38b23138-b637-47d2-bf86-058413e2933a" />

```bash
[2026-04-15 10:49:14 +0200] information/IcingaDB: Pending items queue size: 94, Redis queries: 0, oldest item age: 302ms
[2026-04-15 10:50:09 +0200] information/IcingaDB: Pending queries: 0 (Input: 219/s; Output: 219/s)
[2026-04-15 10:54:14 +0200] information/IcingaDB: Pending items queue size: 0, Redis queries: 0
[2026-04-15 10:55:09 +0200] information/IcingaDB: Pending queries: 0 (Input: 5/s; Output: 5/s)
[2026-04-15 10:59:14 +0200] information/IcingaDB: Pending items queue size: 1, Redis queries: 0, oldest item age: 305ms
```

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -53,10 +54,32 @@ void IcingaDB::PendingItemsThreadProc()
+		// Don't spam the logs when the items are processed in a timely manner, but only if there are
+		// items pending for a while (i.e. the Redis connection might be saturated or down, or there
+		// might be a sudden burst of changes that need to be processed).
+		if (lastLogTime + logInterval <= now || (!seqView.empty() && seqView.front().EnqueueTime + minItemAge < now - 5s)) {
```
> Don't spam the logs when the items are processed in a timely manner

But otherwise, do spam the logs? I mean wouldn't that, once you made it over the threshold, log this for every single item?

What about choosing `logInterval` based on the age of the oldest item? Something like increase to logging every 10 seconds if there are items older than 5 seconds.

yhabteab replied:
> I mean wouldn't that, once you made it over the threshold, log this for every single item?

Yep, I missed that. Fixed now, but I'm not sure if that's exactly what you're suggesting with your comment above, so pls have a look at it.

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -417,3 +440,28 @@ void IcingaDB::EnqueueRelationsDeletion(const String& id, icingadb::task_queue::
+			oldestItemAge = ch::steady_clock::now() - seqView.front().EnqueueTime;
+		}
+	}
+	perfdata->Add(new PerfdataValue("icinga2_pending_items_count", static_cast<double>(queueSize), true));
```
That `true` means `counter=true` which isn't the case here (count ≠ counter).

> The UoM “c” represents a continuous counter (e.g. interface traffic counters).

That's supposed to mean that the value only increases (and only resets back to 0 if the source loses its state).

yhabteab replied:
I mean, the size does go back to 0 at some point whether or not the checkcommand gets that state to see is another question, but I'll drop it.

julianbrost replied:
I think our documentation (which seems to be heavily inspired by the [monitoring plugins development guidelines](https://www.monitoring-plugins.org/doc/guidelines.html#AEN197)) isn't that clear, but I'm pretty sure it's supposed to mean the same as [counter as a metric type in Prometheus](https://prometheus.io/docs/concepts/metric_types/#counter) (in contrast to gauge).

yhabteab replied:
Dropped.

julianbrost reviewed (commented):
> > A warning message in case this is way behind would be helpful as well.
> 
> Why warning? I've just added a normal (info) logging at regular interval (always every 5m unless the worker is lagging behind)

In that you've added a distinction based on whether the oldest item is due for more than 5 seconds. Isn't that because that usually shouldn't happen, so if it does, there's a good chance that something is wrong here? So wouldn't it make sense to also increase the log severity to warning in that case? So that if something would go wrong and you look at the logs, you have a chance of finding the cause by looking at critical/warning messages first, instead of having to know which information messages you need to look for.

Also, the naming "pending items" it pretty generic, something a bit more descriptive would be nice. Maybe include something like background update queue?

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,471 @@
+	// The regular interval at which to log the statistics about the pending items queue.
+	constexpr ch::seconds logInterval{5min};
+	// The interval at which to log the statistics about the pending items when the worker is under heavy load.
+	constexpr ch::seconds heavyLoadLogInterval{10};
```
I don't see a reason to mix units here like you do. They are only used for comparing with `elapsed`, which is in nanoseconds (on my machine at least), so it's not like everything is consistently in seconds so far. This, why not do it like that:

```suggestion
	// The regular interval at which to log the statistics about the pending items queue.
	constexpr auto logInterval{5min};
	// The interval at which to log the statistics about the pending items when the worker is under heavy load.
	constexpr auto heavyLoadLogInterval{10s};
```

Ideally, also make sure that `minItemAgo` is also consistent with that.

yhabteab replied:
That changes absolutely nothing, the variables still have different units but fine, changed them anyway.

julianbrost replied:
That wasn't my point, my point was that you set a `std::chrono::seconds` from a minute literal (`5min`), but there's no reason why to store this in type seconds.

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,471 @@
+		// make the user aware of the situation as that's usually an indication of something being wrong.
+		if (elapsed >= logInterval || (laggingBehind && elapsed >= heavyLoadLogInterval)) {
+			lastLogTime = now;
+			Log log(elapsed < logInterval ? LogWarning : LogInformation, "IcingaDB");
```
Why are you checking `elapsed < logInterval` to determine the log level? Wouldn't checking `laggingBehind` have the same effect?

yhabteab replied:
Changed.

julianbrost commented on lib/icingadb/icingadb-worker.cpp:
```diff
@@ -0,0 +1,471 @@
+		}
+	}
+	perfdata->Add(new PerfdataValue("icinga2_pending_items_count", static_cast<double>(queueSize)));
+	perfdata->Add(new PerfdataValue("icinga2_pending_items_backlog", ch::duration_cast<ch::duration<double>>(oldestItemAge).count(), false, "seconds"));
```
Having a more descriptive name for the perfdata labels would be nice as well.

yhabteab replied:
Changed.

yhabteab committed:
Inline `SendNextUpdate` & remove superfluous `m_RconWorker` checks

yhabteab commented:
Sorry, just pushed something new again - I've now inlined the `SendNextUpdate` method since the it's used by the worker exclusively now. I could've inlined the `UpdateState` method too but it's somewaht a bit bigger and also contain a doc block, so I didn't want to remove that (and potentially start a long running discussion again). Also, I've removed some superfluous checks on the `m_RconWorker` pointer.

julianbrost reviewed (approved):
🎉

julianbrost commented:
@jschmidt-icinga You still have an unresolved conversation (https://github.com/Icinga/icinga2/pull/10619#discussion_r2513335068). I think it's long outdated, do you want to have a look at it (that conversation or the whole PR) nonetheless?

jschmidt-icinga commented:
>do you want to have a look at it (that conversation or the whole PR) nonetheless?

I haven't looked at it in a while, and I'm not that experienced in the Redis protocol, so a thorough review would take at least a few days up to a week and I don't think we have the time for that. Therefore I'd skip this one entirely, unless you want me to take a quick look at the code quality aspects only.

julianbrost commented:
That was just meant as a friendly "your last chance step in before I merge it" reminder, so not a request from my side.
